src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs

```diff
@@ -0,0 +1,356 @@
+        }
+
+        private bool VisitSimpleType()
+        {
+            if (_modelState.HasReachedMaxErrors)
+            {
+                SuppressValidation(_key);
+                return false;
+            }
+
+            return ValidateNode();
+        }
+
+        private bool VisitChildren(IValidationStrategy strategy)
+        {
+            var isValid = true;
+            var enumerator = strategy.GetChildren(_metadata, _key, _model);
+
+            while (enumerator.MoveNext())
+            {
+                var metadata = enumerator.Current.Metadata;
+                var model = enumerator.Current.Model;
+                var key = enumerator.Current.Key;
+
+                isValid &= Visit(metadata, key, model);
+            }
+
+            return isValid;
+        }
+
+        private IList<IModelValidator> GetValidators(ModelMetadata metadata)
+        {
+            var context = new ModelValidatorProviderContext(metadata);
+            _validatorProvider.GetValidators(context);
+            return context.Validators.OrderBy(v => v, ValidatorOrderComparer.Instance).ToList();
+        }
+
+        private void SuppressValidation(string key)
+        {
+            var entries = _modelState.FindKeysWithPrefix(key);
+            foreach (var entry in entries)
+            {
+                entry.Value.ValidationState = ModelValidationState.Skipped;
+            }
+        }
+
+        private bool ShouldValidateProperties(ModelMetadata metadata)
+        {
+            var count = _excludeFilters.Count;
+            for (var i = 0; i < _excludeFilters.Count; i++)
+            {
+                if (_excludeFilters[i].IsTypeExcluded(metadata.UnderlyingOrModelType))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private ValidationStateEntry GetValidationEntry(object model)
+        {
+            if (model == null || _validationState == null)
+            {
+                return null;
+            }
+
+            ValidationStateEntry entry;
+            _validationState.TryGetValue(model, out entry);
+            return entry;
+        }
+
+        private struct StateManager : IDisposable
+        {
+            private readonly ValidationVisitor _visitor;
+            private readonly object _container;
+            private readonly string _key;
+            private readonly ModelMetadata _metadata;
+            private readonly object _model;
+            private readonly object _newModel;
+            private readonly IValidationStrategy _strategy;
+
+            public static StateManager Recurse(
+                ValidationVisitor visitor,
+                string key,
+                ModelMetadata metadata,
+                object model,
+                IValidationStrategy strategy)
+            {
+                var recursifier = new StateManager(visitor, model);
+
+                visitor._container = visitor._model;
```

Comment:
Container is what we're inside. If we're validating a property, then model is the object we evaluated the property from.